# Variant Annotation - Python
This scripts reads an output file generated by SnpEff, normalizes and corrects most of the consistent errors of SnpEff.
The input file that is read by the script consists of two columns. 
The first column is the variant id. Its type is an integer.
The second column is the "ANN" column output generated by SnpEff. 
The bash command used to generate that file is the following:

```
snpEff ann -config snpEff.config -upDownStreamLen 2000 -spliceSiteSize 0 -spliceRegionExonSize 0 -spliceRegionIntronMax 0 -spliceRegionIntronMin 0 NC_000962.3 tmp.vcf | grep -v "^#" | cut -f3,8 > annotated_file.txt

```

# Solo algorithm - R implementation
Contains all scripts necessary to run the SOLO algorithm in R.

# Solo algorithm - Stata implementation
Contains all scripts necessary to run the SOLO algorithm in Stata.

# Input data
Input data. 
## full_genotypes
Contains each variant (i.e. features of the models) for each sample. The data is arranged in two sub-levels of folders, one for drugs and one for tiers. The csv files contain 6 columns.

```
sample_id,resolved_symbol,variant_category,predicted_effect,neutral,"max(af)",position
```

- resolved_symbol is the gene name to which the variant belongs to
- variant_category is the final variant name. it is the result of the categorization algorithm which can be found in the PySpark folder.

## phenotype
```
sample_id,phenotypic_category,phenotype,box
```

- phenotypic_category is the overall category the test was assigned to ("WHO" or "ALL")
- phenotype is the result (R or S)
- box is the subcategory of the test. Refer to the mutation catalogue output for more explanations